### ISSUE TITLE
Calcium dynamics

### ISSUE THREAD

**Issue by SMPugliese** [jonescompneurolab/hnn-core#111]
As many of you know I have spent the last year or so changing the model L5 pyramidal cells so that the simulated voltage traces in the dendrites look more similar to electrophysiological data recorded from pyramidal neuron dendrites, especially during calcium spikes/bursts. To do this, I changed the L5_pyramidal.py so that the dendritic Na, Kv, and Ca (L-type) conductances are distributed as functions of the distance from the soma, rather than setting the whole dendrite equal to one value. Here is the modified file for anyone who wants to look at or use these (saved as .txt because github didn't allow attaching .py files): [L5_pyramidal.txt](https://github.com/jonescompneurolab/hnn-core/files/4914122/L5_pyramidal.txt)

### For anyone using it  
**Important**: I believe that if you are only changing this file out for the existing L5_pyramidal.py file you will run into an error with __create_all_IClamp() that will stop the code from working. I will get around to fixing this soon but in the meantime just copy the version of this function that's in the original code and replace this function. The reason is that I added functionality to inject current into the dendrite for some of my simulations but that involved changing other files as well. Sorry about this!

Within the file I have included a way to switch back and forth from these distributions ("updated_HNN_dends") to the regular HNN distributions ("classic_HNN_dends") which are constant values for everything except HCN channels. Just comment one or the other out under the _biophys_dends function like so
>def __biophys_dends(self):
    # self.classic_HNN_dends()
    self.updated_HNN_dends()

You have to save the file, then **quit HNN and restart it** in order for this to do anything because this code gets run right when it is launched.

Right now the dendritic and somatic conductances are still adjustable from the GUI or the param files. The parameters I've been using for these three channels (Cell Parameters > L5 Pyr Biophysics) are:

> L5Pyr_soma_gkbar_hh2: 0.06
L5Pyr_soma_gnabar_hh2: 0.32
L5Pyr_soma_gbar_ca: 10.0
L5Pyr_dend_gkbar_hh2: 0.0001
L5Pyr_dend_gnabar_hh2: 0.0028
L5Pyr_dend_gbar_ca: 40.0

The dendritic conductances are now functions of these values. Some of these changes are pretty drastic. With these values, the channel density distribution will now look like this (dotted lines are "classic" conductances).
![fig_2_5 copy 2](https://user-images.githubusercontent.com/32277133/87336670-38202200-c510-11ea-99a9-46a499516685.png)

**Files**
I have roughly hand-tuned the evoked input parameters for the threshold and suprathreshold ERP parameter files so that, with these changes to the L5 pyramidal cells, the output provides a good fit to the data. These need to be improved/optimized more. Once again I had to upload as .txt but they should be .param files.
![image](https://user-images.githubusercontent.com/32277133/87337233-25f2b380-c511-11ea-8ebb-7a105afb8bc1.png)
[ERP_yes.txt](https://github.com/jonescompneurolab/hnn-core/files/4914267/ERP_yes.txt)

![image](https://user-images.githubusercontent.com/32277133/87337353-4b7fbd00-c511-11ea-91cf-b637d60665c0.png)
[ERP_suprathreshold.txt](https://github.com/jonescompneurolab/hnn-core/files/4914269/ERP_suprathreshold.txt)

### For HNN development
I am sure there is a much better way for this code to be written and organized! Not to mention that hnn-core gets rid of the L5_pyramidal.py file. Some things to think about include 

- whether there would be a way for users to easily switch back and forth from the "classic" and "updated" distributions from the GUI, rather than the bad solution I have of changing the source code and relaunching
- whether the parameters I have characterizing the distributions (e.g. the distance where the calcium conductance levels off) ought to be hard-coded as they are now or adjustable by the user.

Also, I changed and added a lot more files in order to be able to record and plot voltage traces, calcium currents, and calcium conductances in individual compartments of the L5 pyramidal neurons. I think some of these plots could be useful to include in the new versions of HNN. I've just created a public repository SMPugliese/hnn_calcium ([https://github.com/SMPugliese/hnn_calcium](https://github.com/SMPugliese/hnn_calcium)) with the most up to date files I have on my computer. Let me know if there is another place you would prefer me to put these files.

**Comment by jasmainak** [comment]
@SMPugliese thanks for opening the issue! Did you close it intentionally? I'm reopening it to discuss further. Can you edit your comment above to add a link to your repository?

I had created a pull request with the files that Carmen sent me [here](https://github.com/jonescompneurolab/hnn-core/pull/110/files). Obviously this is not everything you have done. In general, I think it's a good idea to contribute back to the HNN repository. That way, you can have your changes up to date with everything else in HNN and you are not working off an older version. It's also a good opportunity to test whatever you change.

One obvious way I see to incorporate your change is to allow a `L5pyr` argument in the `Network` class that is `None` by default. This way you can override the default `L5Pyr` class with one you provide. Something like:

```py
from hnn_core.pyramidal import L5Pyr

class L5PyrCustom(L5Pyr):

    def __init__(self, gid, pos, p):
        L5Pyr.__init__(gid, pos, p)

    def _biophys_dends(self):
        # your modified code here

net = Network(...., L5Pyr=L5PyrCustom)
dpl = simulate_dipole(net)
```

**Comment by SMPugliese** [comment]
Thanks! I've edited to include the link. I can submit a pull request to this repository later as well.

**Comment by jasmainak** [comment]
sounds good, start something and we'll help you if you get stuck!

**Comment by jasmainak** [comment]
closed by #333
